### PR TITLE
Remove JsonEqual

### LINKED ISSUE #2296: Remove JsonEqual
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

All `Array` implementations must implement the `JsonEqual` trait. However:

- UnionArray will always panic
- DictionaryArray is incorrect (#2294)
- It is unclear to me why we would special case JSON
- It complicates making JSON an optional feature

**Describe the solution you'd like**

Rather than maintaining parallel logic to interpret `serde_json::Value` as arrow arrays, I wonder if we can just remove the `JsonEqual` functionality and encourage conversion to arrow arrays. I'm unclear of a use-case that would require highly-performant comparison of arrow and JSON data

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2296.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: integration-testing/src/lib.rs [hunk 1]
```diff
@@ -88,717 +76,3 @@ pub fn read_json_file(json_name: &str) -> Result<ArrowFile> {
...
-    for (field, json_col) in schema.fields().iter().zip(json_batch.columns) {
-        let col = array_from_json(field, json_col, json_dictionaries)?;
-        columns.push(col);
-    }
-
-    RecordBatch::try_new(Arc::new(schema.clone()), columns)
-}
-
-/// Construct an Arrow array from a partially typed JSON column
-fn array_from_json(
```
Comment (tustvold) [inline integration-testing/src/lib.rs:108]: This appears to have just been moved

**Review by tustvold — APPROVED** [review]
What a fast turn around, love it 👍

File: arrow/src/temporal_conversions.rs [hunk 2]
```diff
@@ -127,7 +127,7 @@ pub fn timestamp_ns_to_datetime(v: i64) -> NaiveDateTime {
         // extract seconds from nanoseconds
         v / NANOSECONDS,
         // discard extracted seconds
-        (v % NANOSECONDS) as u32,
+        if v > 0 { (v % NANOSECONDS) as u32 } else { 0 },
```
Comment (viirya) [inline arrow/src/temporal_conversions.rs:130]: When `v` is negative, this parameter causes `invalid or out-of-range datetime` error. Just temporary avoid the failure here. Other function like `timestamp_us_to_datetime` might have same issue.

Comment (viirya) [reply arrow/src/temporal_conversions.rs:130]: #2325

File: arrow/src/util/integration_util.rs [hunk 3]
```diff
@@ -945,22 +1258,25 @@ mod tests {
             .len(3)
             .add_buffer(value_offsets)
             .add_child_data(value_data.into_data())
+            .null_bit_buffer(Some(Buffer::from([0b00000011])))
```
Comment (viirya) [inline arrow/src/util/integration_util.rs:1261]: This appears to be a bug in previous code. The `integration.json` specifies null buffer like the update, but it doesn't reflect that before.

File: arrow/src/util/integration_util.rs [hunk 4]
```diff
@@ -945,22 +1258,25 @@ mod tests {
...
-            ),
+        let struct_data_type = DataType::Struct(vec![
+            Field::new("int32s", DataType::Int32, true),
+            Field::new("utf8s", DataType::Utf8, true),
         ]);
+        let struct_data = ArrayData::builder(struct_data_type)
+            .len(3)
+            .add_child_data(structs_int32s.data().clone())
+            .add_child_data(structs_utf8s.data().clone())
+            .null_bit_buffer(Some(Buffer::from([0b00000011])))
```
Comment (viirya) [inline arrow/src/util/integration_util.rs:1276]: same as to reflect the validity data in `integration.json`.

**Review by tustvold — APPROVED** [review]

**Comment by alamb** [comment]
This is a nice change -- thank you @viirya  and @tustvold

**Comment by viirya** [comment]
Thanks @tustvold @alamb

**Pull request merged**